### PR TITLE
Moved the metadata into setup.cfg

### PR DESCRIPTION
**Pull request by KOLANICH**
Added pyproject.toml
Version is now fetched from git tags.

### CODE REVIEW

**Comment by jstasiak** [comment]
Thank you for the contribution! I see two changes in this PR and I'll give separate opinions on them:

* Moving metadata from `setup.py` and `setup.cfg` – I like it and I'm happy to accept this change if it's submitted alone, with the commit message subject in the imperative ("Move...", not "Moved...") and few more words of explanation *why* this is being done (I can see the reasoning but I want it to be reflected in the repository history). I should really document those quidelines in a PR template.
* Reading the version from git tag with the associated machinery, build-time dependency, `pyproject.toml` etc. – I don't see much benefit here and I'm unlikely to accept this part. I'm happy to learn about the benefits of this approach and the motivation behind this change though (which also need to be included in the commit message if we decide to go through with this).

**Comment by KOLANICH** [comment]
>I don't see much benefit here and I'm unlikely to accept this part. I'm happy to learn about the benefits of this approach and the motivation behind this change though (which also need to be included in the commit message if we decide to go through with this).

>Reading the version from git tag with the associated machinery

The benefit of this approach is following:

1. we use a single source of truth, and it is a git tag, eliminating the possibility of inconsistency, when a git tag is set to one version, and the hardcoded value is another one.
2. `setuptools_scm` for non-tag commits also stores part of commithash as a part of version. Then this part of commit hash can be retrieved, that provides additional information in the case of troubleshooting.

The benefit of using a reusable plugin to setuptools are

1. we can then eliminate `setup.py` entirely (yes, I have kept it, but in fact it is completely not needed). `setup.py` is itself a security risk (yes, I'm a paranoid, and it is actually a good thing ;), and yes, this already have been exploited in the wild), especially when pip is run from root, it was one of the main reasons behind declarative metadata with rigid schema. By using a plugin we shift the trust from potentially malicious setup.py supplied by a package maintainer to a well-known plugin used by lot of packages and checked by many people and supplied by a third party other than the maintainer of the package being installed.
2. better code reuse

The benefits of `pyproject.toml` are that it allows the package building be initiated by a PEP 517 build frontend, such as [`build`](https://github.com/pypa/build), which is a recommended and unified (any build backend supporting pep517 (setuptools, flit, poetry) can be executed by a simple and the same command `python3 -m build -nwx .` (and in case of `setup.cfg`-only projects, the only way is PEP 517 frontend, since there is no setup.py one can execute), which is very convenient especially for automated scripts, like CI) tool for it.

**Comment by KOLANICH** [comment]
If you don't want to fetch version from tags, I can replace `setuptools_scm` with `read_version`, which is another setuptools plugin, which does the same you did with your own custom version carving, but instead it works on AST level and also reusable and also enables us to eliminate `setup.py`.

But IMHO fetching the version from git tags is more beneficial.

**Comment by KOLANICH** [comment]
I have separated the change other than setup.cfg into other branches

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -17,8 +17,20 @@ jobs:
...
       run: |
+        pip install --upgrade pip setuptools wheel
         pip install --upgrade -r requirements.txt
-        pip install .
+    - name: Build package
+      run: |
+        python ./setup.py bdist_wheel
+        mv ./dist/*.whl ./dist/netaddr-0.CI-py2.py3-none-any.whl
+    - name: Upload the wheel artifact
+      uses: actions/upload-artifact@v2
```
Comment (jstasiak) [inline .github/workflows/ci.yml:27]: Uploading artifacts is out of scope of this PR I'd argue.

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -17,8 +17,20 @@ jobs:
...
+      with:
+        name: wheel
+        path: ./dist/netaddr-0.CI-py2.py3-none-any.whl
+    - name: Install package
+      run: |
+        pip install ./dist/netaddr-0.CI-py2.py3-none-any.whl
```
Comment (jstasiak) [inline .github/workflows/ci.yml:33]: Is any of the changes in this file necessary? It seems to me that the answer is no and those are out of scope but I may be missing something.

**Review by jstasiak — COMMENTED** [review]
Thank you, looks much better. Just two questions, see them inline.

File: .github/workflows/ci.yml (hunk 2 above)
Comment (KOLANICH) [reply .github/workflows/ci.yml:33]: * `setuptools` available by default in the env is damn old (22) and seems to be unable to update itself even if specified (for setup.cfg 30.3.0 (dated to December 2016) is required).
* explicit building wheel is needed for `setuptools_scm`-introducing PR - while the code is installed, it is still imported from the relative dir. `setuptools_scm` generates a file with version, but when `pip install .`, that file goes to a temp dir and then into wheel. When building explicitly the file goes to the repative dir.

File: .github/workflows/ci.yml (hunk 1 above)
Comment (KOLANICH) [reply .github/workflows/ci.yml:27]: uploaded artifacts are nice for debugging, also allow people to use your wheels without building them themselves (though it is discouraged ). But yes, they are out of scope of this PR.

**Comment by jstasiak** [comment]
Hey again,

Unfortunately the patch contains some bits that I'd not want merged (mentioned above) so I have to close this PR.

Thank you for the contribution though – much appreciated.

You gave me inspiration to convert the package directly to `pyproject.toml` which I'm about to do.

**Pull request closed without merging**